MODELIX-584 Update roadmap (#27)
modelix/modelix.docs: docs/modules/ROOT/pages/main/roadmap.adoc

= Roadmap

The following image shows a dense overview of our current roadmap.

image::2024_01_roadmap.png[]

== Key priorities overview

The aim of this roadmap is to provide you with a more detailed overview of our plans.
We have compiled a catalog of our significant areas, which are our top priorities we focus on delivering.
In the following we shortly list these key priorities.
Below you can find more detailed explanations for each of these.

// ------------- 24.1
* *ModelQL*: Dynamic query language for on-demand loading overcoming issues associated with upfront data retrieval and optimizing performance
* *MPS 2023 / 2024 Support*: Provide support in modelix for newer MPS versions
* *LIonWeb* (estimated 24.1): Integration with other web components via a new LionWeb bulk API adapter
// ------------- 24.2

> * *Model-server services*: Streamlined interfaces and structure for services working on model data, providing for example model analysis, checking, and transformation

Reviewer: "example model analysis" is a bit imprecise. Also, the word "service" should probably be defined.

> * *Meta-model in model-server*: A way to store the meta-model of corresponding model content in the model-server (foundation for Language Evolution).

Reviewer: ```suggestion
* *Meta-model in model-server*: A way to store the meta-model of corresponding model content in the model-server (foundation for language evolution).
```

and also for other things? So maybe add "e.g." or something similar.

// ------------- 25.1
* *GraphQL interface*: An endpoint for smaller web applications, leveraging the available tooling ecosystem
* *Projectional text editors in browser*: Projectional text editors for browsers, using Kotlin multiplatform to edit models directly or through server-side rendering
// ------------- 25.2

> * *Graphical Editors*: We plan to provide the data synchronization and the generation of a TypeScript model API to enable graphical editors on the wev.

Reviewer 2: ```suggestion
* *Graphical Editors*: We plan to provide the data synchronization and the generation of a TypeScript model API to enable graphical editors on the web.
```

> * *Graphical Editors*: We plan to provide the data synchronization and the generation of a TypeScript model API to enable graphical editors on the wev.

Reviewer 3: ```suggestion
* *Graphical Editors*: We plan to provide the data synchronization and the generation of a TypeScript model API to enable graphical editors on the web.
```

// ------------- 24.1
[#modelql,reftext="ModelQL"]
=== ModelQL (estimated 24.1)

With the modelix 23.2 release we https://modelix.org/blog/2023/11/02/modelix-platform-release-23.2/#modelql[introduced ModelQL]:
A dynamic way of loading parts of the model on demand while reducing the number of requests to a minimum.
To learn more, head over to the https://docs.modelix.org/modelix/main/core/explanation/modelql.html[ModelQL documentation] and the corresponding https://docs.modelix.org/modelix/main/core/howto/modelql.html[How-Tos].
While the early version allowed us to provide a quick win for large models, we were not able to finish this feature fully.

> Especially https://docs.modelix.org/modelix/main/core/howto/modelql-writing.html[writing support] currently needs some further work.

Reviewer 4: I think the writing support is pretty much done for now.
The functionality of modelql is there, but we should work on stuff like the debugging experience.

Reviewer 2: You could also say we are working on making it more mature/stabalizing it (aka. bugfixing and peformance)

We also want to integrate ModelQL in our https://github.com/modelix/modelix-samples[samples project] to present a graspable example beyond the documentation.

=== MPS 2023 / 2024 Support (estimated 24.1)

In a larger effort to become MPS version independent, we prepare our plugins and interfaces with MPS to provide preliminary support for MPS 2023 for the modelix 24.1 release.
This effort is needed to provide adequate support today, but in the following release 24.2 we plan to fully transition away from dedicated MPS plugins over to https://plugins.jetbrains.com/docs/intellij/intellij-platform.html?from=IJPluginTemplate[IntelliJ Platform] plugins.

> IntelliJ Platform plugins do not require to be written in MPS and allow to be created using Kotlin - just like everything else in modelix.

Reviewer 3: ```suggestion
IntelliJ Platform plugins do not require to be written in MPS and are allowed to be created using Kotlin - just like everything else in modelix.
```

All MPS interfacing parts will use the comparably stable APIs provided by MPS to query and modify MPS models, such as the https://www.jetbrains.com/help/mps/open-api-accessing-models-from-code.html:[Open API] and the https://www.jetbrains.com/help/mps/smodel-language.html:[SModel].
As a result, we do not have to migrate and maintain MPS plugins for multiple MPS versions at the same time.
Instead, we have a single Kotlin codebase which can provide the same functionality to all MPS versions.
This effort will reduce complexity and potential for bugs and allow us to better test and validate our MPS interfacing plugins.

=== LIonWeb (estimated 24.1)
Modelix focuses on a good integration with MPS.

> But if you are interested in the integration of other modeling technologies then check out the LIonWeb project (short for Language Interfaces on the web).

Reviewer 2: A link to the LIonWeb project could be nice.

They specify more general interfaces and use the modelix `model-server` for data storage and exchange in their current prototypes.
For 24.1 we want to extend the modelix `model-server` to support the initial LionWeb bulk protocol for model data storage and retrieval.
The modelix team co-authored and is involved with the LIonWeb initiative.

// ------------- 24.2
[#services, reftext="model-server services"]
=== Model-server services (estimated 24.2)
Modelix already allows you to access model data from any web application, either untyped (i.e. on the M3 layer), or typed (i.e. on the M2 layer or domain layer).
As soon as you have model content in a web application the question regarding correct scoping, typing, and other checks arises.
In MPS we grew accustomed to easy integration of these different language aspects.
In modelix we envision `model-server` *services* to do this type of model analysis, checking, and eventually transformation.
In fact, this can already be done today:
Any application accessing model data from a `model-server` can provide external APIs for the above-mentioned features.

> However, we plan to streamline interfaces and approaches used by efforts that typical services can provide.

Reviewer: "efforts" is a bit wordy

As a consequence, we will need meta-model information in the `model-server` and plan to add this as a foundation for Language Evolution.

Until now, there was no need for this knowledge to be available directly but technically the meta-model is already stored, though implicitly encoded in the storage, but the `model-server` does not use or interpret this knowledge.
We have some ideas we need to further refine and explore in the following releases.

// ------------- 25.1
=== GraphQL interface (estimated 25.1)
As with all other parts of modelix, we prefer using established and state-of-the-art tools.
Consequently, for smaller web applications we consider to provide a simple https://graphql.org/[GraphQL] endpoint, which can provide immediate benefit.
Especially the access to a powerful tooling ecosystem with GUIs, editor integrations, code generation, linting, analytics, and more is a direct win.

However, it is important to note, that this approach will not scale:
As soon as the model size grows or the strong integration requirements exist, other tools such as our <<modelql>> or the <<services>> are needed.

=== Projectional text editors in browser (estimated 25.1)

> Similar to our previously developed https://github.com/modelix/modelix/tree/mps/2020.3/mps/org.modelix.notation[notation language] prototype (based on the old shadow models), we want to provide

Reviewer 2: > Similar to our previously developed notation language prototype, we want to provide

The sentence is incomplete.

A default notation is used for concepts without a web specific notation, just as MPS does it.

> We envision projectional text editor components implemented in Kotlin multiplatform for editing models in the browser, either directly or via server side rendering.

Reviewer 3: ```suggestion
We envision projectional text editor components implemented in Kotlin multiplatform for editing models in the browser, either directly or via server-side rendering.
```
